config.go

```diff
@@ -0,0 +1,288 @@
+	TestType        string `env:"test_type,opt[instrumentation,robo,gameloop]"`
+	TestDevicesList string `env:"test_devices"`
+	TestDevices     []*testing.AndroidDevice
+	AppPackageID    string `env:"app_package_id"`
+
+	// test setup
+	AutoGoogleLogin          bool   `env:"auto_google_login,opt[true,false]"`
+	EnvironmentVariablesList string `env:"environment_variables"`
+	EnvironmentVariables     []*testing.EnvironmentVariable
+	ObbFilesList             string `env:"obb_files_list"`
+	ObbFiles                 []string
+
+	// shared debug
+	TestTimeout           float64 `env:"test_timeout,required"`
+	FlakyTestAttempts     int     `env:"num_flaky_test_attempts,range[0..10]"`
+	DownloadTestResults   bool    `env:"download_test_results,opt[true,false]"`
+	DirectoriesToPullList string  `env:"directories_to_pull"`
+	DirectoriesToPull     []string
+	VerboseLog            bool `env:"use_verbose_log,opt[true,false]"`
+
+	// instrumentation
+	InstTestPackageID   string `env:"inst_test_package_id"`
+	InstTestRunnerClass string `env:"inst_test_runner_class"`
+	InstTestTargets     string `env:"inst_test_targets"`
+	UseOrchestrator     bool   `env:"inst_use_orchestrator,opt[true,false]"`
+
+	// robo
+	RoboInitialActivity string `env:"robo_initial_activity"`
+	RoboDirectives      string `env:"robo_directives"`
+	RoboScenarioFile    string `env:"robo_scenario_file"`
+	RoboMaxDepth        string `env:"robo_max_depth"`
+	RoboMaxSteps        string `env:"robo_max_steps"`
+
+	// loop
+	LoopScenarios       string `env:"loop_scenarios"`
+	LoopScenarioLabels  string `env:"loop_scenario_labels"`
+	LoopScenarioNumbers string `env:"loop_scenario_numbers"`
+
+	// deprecated
+	ApkPath string `env:"apk_path"`
+}
+
+func (configs *ConfigsModel) print() {
+	log.Infof("Configs:")
+	log.Printf("- AppPath: %s", configs.AppPath)
+	if configs.ApkPath != "" {
+		log.Printf("- ApkPath: %s", configs.ApkPath)
+	}
+	if configs.AppPackageID != "" {
+		log.Printf("- AppPackageID: %s", configs.AppPackageID)
+	}
+	log.Printf("- TestTimeout: %f", configs.TestTimeout)
+	log.Printf("- FlakyTestAttempts: %d", configs.FlakyTestAttempts)
+	log.Printf("- DownloadTestResults: %t", configs.DownloadTestResults)
+	log.Printf("- DirectoriesToPull: %s", configs.DirectoriesToPullList)
+	log.Printf("- AutoGoogleLogin: %t", configs.AutoGoogleLogin)
+	log.Printf("- EnvironmentVariables: %s", configs.EnvironmentVariablesList)
+	log.Printf("- ObbFilesList: %s", configs.ObbFilesList)
+	log.Printf("- TestDevices:\n---")
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, 3, ' ', 0)
+	if _, err := fmt.Fprintln(w, "Model\tAPI Level\tLocale\tOrientation\t"); err != nil {
+		failf("Failed to write in tabwriter, error: %s", err)
+	}
+	scanner := bufio.NewScanner(strings.NewReader(configs.TestDevicesList))
+	for scanner.Scan() {
+		device := scanner.Text()
+		device = strings.TrimSpace(device)
+		if device == "" {
+			continue
+		}
+
+		deviceParams := strings.Split(device, ",")
+
+		if len(deviceParams) != 4 {
+			continue
```

Comment:
Note: it would be helpful to add a debug message, that this device was ignored.

Reply:
Changed to log test devices after they were parsed.